File: razarion-ui-service/src/main/java/com/btxtech/uiservice/cockpit/item/ItemCockpitService.java
package com.btxtech.uiservice.cockpit.item;


import com.btxtech.shared.CommonUrl;
import com.btxtech.shared.dto.BaseItemPlacerConfig;
import com.btxtech.shared.gameengine.ItemTypeService;
import com.btxtech.shared.gameengine.datatypes.itemtype.BaseItemType;
import com.btxtech.shared.gameengine.datatypes.itemtype.ItemType;
import com.btxtech.shared.gameengine.datatypes.workerdto.PlayerBaseDto;
import com.btxtech.shared.gameengine.datatypes.workerdto.SyncBaseItemSimpleDto;
import com.btxtech.shared.gameengine.datatypes.workerdto.SyncBoxItemSimpleDto;
import com.btxtech.shared.gameengine.datatypes.workerdto.SyncItemSimpleDto;
import com.btxtech.shared.gameengine.datatypes.workerdto.SyncResourceItemSimpleDto;
import com.btxtech.shared.utils.CollectionUtils;
import com.btxtech.uiservice.Group;
import com.btxtech.uiservice.SelectionEvent;
import com.btxtech.uiservice.SelectionEventService;
import com.btxtech.uiservice.SelectionService;
import com.btxtech.uiservice.control.GameEngineControl;
import com.btxtech.uiservice.control.GameUiControl;
import com.btxtech.uiservice.item.BaseItemUiService;
import com.btxtech.uiservice.item.SyncBaseItemMonitor;
import com.btxtech.uiservice.item.SyncItemMonitor;
import com.btxtech.uiservice.itemplacer.BaseItemPlacerService;

import jakarta.inject.Inject;
import jakarta.inject.Provider;
import jakarta.inject.Singleton;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

@Singleton
public class ItemCockpitService {
    private final Logger logger = Logger.getLogger(ItemCockpitService.class.getName());
    private final ItemTypeService itemTypeService;
    private final Provider<BaseItemUiService> baseItemUiService;
    private final Provider<GameUiControl> gameUiControl;
    private final BaseItemPlacerService baseItemPlacerService;
    private final Provider<GameEngineControl> gameEngineControl;
    private final SelectionService selectionService;
    private final Collection<BuildupItemCockpit> buildupItemCockpits = new ArrayList<>();
    private ItemCockpitFrontend itemCockpitFrontend;

    @Inject
    public ItemCockpitService(SelectionService selectionService,
                              Provider<GameEngineControl> gameEngineControl,
                              BaseItemPlacerService baseItemPlacerService,
                              Provider<GameUiControl> gameUiControl,
                              Provider<BaseItemUiService> baseItemUiService,
                              ItemTypeService itemTypeService,
                              SelectionEventService selectionEventService) {
        this.selectionService = selectionService;
        this.gameEngineControl = gameEngineControl;
        this.baseItemPlacerService = baseItemPlacerService;
        this.gameUiControl = gameUiControl;
        this.baseItemUiService = baseItemUiService;
        this.itemTypeService = itemTypeService;
        selectionEventService.receiveSelectionEvent(this::onOwnSelectionChanged);
    }

    public void init(ItemCockpitFrontend itemCockpitFrontend) {
        this.itemCockpitFrontend = itemCockpitFrontend;
    }

    private void onOwnSelectionChanged(SelectionEvent selectionEvent) {
        switch (selectionEvent.getType()) {
            case CLEAR:
                buildupItemCockpits.forEach(BuildupItemCockpit::releaseMonitor); // TODO releaseMonitor
                buildupItemCockpits.clear();
                itemCockpitFrontend.dispose();
                break;
            case OWN:
                Group selectedGroup = selectionEvent.getSelectedGroup();
                Map<BaseItemType, Collection<SyncBaseItemSimpleDto>> itemTypes = selectedGroup.getGroupedItems();
                if (selectedGroup.getCount() == 1) {
                    itemCockpitFrontend.displayOwnSingleType(1, createOwnItemCockpit(CollectionUtils.getFirst(itemTypes.keySet()), selectedGroup));
                } else {
                    if (itemTypes.size() == 1) {
                        itemCockpitFrontend.displayOwnSingleType(itemTypes.size(), createOwnItemCockpit(CollectionUtils.getFirst(itemTypes.keySet()), selectedGroup));
                    } else {
                        itemCockpitFrontend.displayOwnMultipleItemTypes(createOwnMultipleInfo(itemTypes));
                    }
                }
                break;
            case OTHER:
                itemCockpitFrontend.displayOtherItemType(createOtherInfo(selectionEvent.getSelectedOther()));
                break;
        }
    }

    private OwnItemCockpit createOwnItemCockpit(BaseItemType baseItemType, Group selectedGroup) {
        OwnItemCockpit ownInfoPanel = createSimpleOwnItemCockpit(baseItemType, selectedGroup.getItems());

        ownInfoPanel.buildupItemInfos = createBuildupItemInfos(baseItemType, selectedGroup);

        ownInfoPanel.itemContainerInfo = createItemContainerInfo(baseItemType, selectedGroup);

        return ownInfoPanel;
    }

    private OwnMultipleIteCockpit[] createOwnMultipleInfo(Map<BaseItemType, Collection<SyncBaseItemSimpleDto>> itemTypes) {
        return itemTypes.entrySet().stream().map(entry -> {
            OwnMultipleIteCockpit ownMultipleItemInfo = new OwnMultipleIteCockpit() {
                @Override
                public void onSelect() {
                    try {
                        selectionService.keepOnlyOwnOfType(entry.getKey());
                    } catch (Throwable t) {
                        logger.log(Level.WARNING, t.getMessage(), t);
                    }
                }
            };
            ownMultipleItemInfo.count = entry.getValue().size();
            ownMultipleItemInfo.ownItemCockpit = createSimpleOwnItemCockpit(entry.getKey(), entry.getValue());
            return ownMultipleItemInfo;
        }).toArray(OwnMultipleIteCockpit[]::new);
    }

    private OtherItemCockpit createOtherInfo(SyncItemSimpleDto otherSelection) {
        OtherItemCockpit otherInfoPanel = new OtherItemCockpit();
        otherInfoPanel.id = otherSelection.getId();
        ItemType itemType = null;
        if (otherSelection instanceof SyncBaseItemSimpleDto) {
            itemType = itemTypeService.getBaseItemType(otherSelection.getItemTypeId());
            SyncBaseItemSimpleDto syncBaseItem = (SyncBaseItemSimpleDto) otherSelection;
            PlayerBaseDto base = baseItemUiService.get().getBase(syncBaseItem.getBaseId());
            otherInfoPanel.baseId = base.getBaseId();
            switch (base.getCharacter()) {
                case HUMAN:
                    if (base.getUserId() == null) {
                        otherInfoPanel.baseName = "Unregistered user";
                    } else if (base.getName() == null || base.getName().trim().isEmpty()) {
                        otherInfoPanel.baseName = "Unnamed user";
                    } else {
                        otherInfoPanel.baseName = base.getName();
                    }
                    otherInfoPanel.friend = true;
                    otherInfoPanel.bot = false;
                    break;
                case BOT:
                    otherInfoPanel.baseName = base.getName();
                    otherInfoPanel.bot = true;
                    break;
                case BOT_NCP:
                    otherInfoPanel.baseName = base.getName();
                    otherInfoPanel.friend = true;
                    otherInfoPanel.bot = true;
                    break;
                default:
                    throw new UnsupportedOperationException("Unknown character: " + base.getCharacter());
            }
        } else if (otherSelection instanceof SyncResourceItemSimpleDto) {
            itemType = itemTypeService.getResourceItemType(otherSelection.getItemTypeId());
            otherInfoPanel.resource = true;
        } else if (otherSelection instanceof SyncBoxItemSimpleDto) {
            itemType = itemTypeService.getBoxItemType(otherSelection.getItemTypeId());
            otherInfoPanel.box = true;
        }
        if (itemType != null) {
            otherInfoPanel.imageUrl = CommonUrl.getImageServiceUrlSafe(itemType.getThumbnail());
            otherInfoPanel.itemTypeName = itemType.getName();
            otherInfoPanel.itemTypeDescr = itemType.getDescription();
        }
        return otherInfoPanel;
    }

    private BuildupItemCockpit[] createBuildupItemInfos(BaseItemType baseItemType, Group selectedGroup) {
        List<Integer> ableToBuildIds;
        Consumer<BaseItemType> onBuildCallback;
        if (baseItemType.getBuilderType() != null) {
            ableToBuildIds = baseItemType.getBuilderType().getAbleToBuildIds();
            onBuildCallback = (itemType) -> {
                BaseItemPlacerConfig baseItemPlacerConfig = new BaseItemPlacerConfig().baseItemCount(1).baseItemTypeId(itemType.getId());
                baseItemPlacerService.activate(baseItemPlacerConfig, true, decimalPositions -> {
                    gameEngineControl.get().buildCmd(selectedGroup.getItems().stream().findFirst().orElseThrow(RuntimeException::new), CollectionUtils.getFirst(decimalPositions), itemType);
                });
            };
        } else if (baseItemType.getFactoryType() != null) {
            ableToBuildIds = baseItemType.getFactoryType().getAbleToBuildIds();
            // Factory
            onBuildCallback = (itemType) -> {
                selectedGroup.getSyncBaseItemsMonitors().stream()
                        .filter(syncBaseItemMonitor -> syncBaseItemMonitor.getConstructingBaseItemTypeId() == null)
                        .map(SyncItemMonitor::getSyncItemId)
                        .findFirst()
                        .ifPresent(factoryId -> gameEngineControl.get().fabricateCmd(factoryId, itemType));
            };
        } else {
            return null;
        }

        return ableToBuildIds.stream()
                .filter(itemTypeId -> gameUiControl.get().getPlanetConfig().imitation4ItemType(itemTypeId) > 0)
                .map(itemTypeId -> {
                    BaseItemType itemType = itemTypeService.getBaseItemType(itemTypeId);
                    BuildupItemCockpit buildupItemInfo = new BuildupItemCockpit() {
                        @Override
                        public void onBuild() {
                            try {
                                onBuildCallback.accept(itemType);
                            } catch (Throwable t) {
                                logger.log(Level.WARNING, t.getMessage(), t);
                            }
                        }

                        @Override
                        public void setAngularZoneRunner(AngularZoneRunner angularZoneRunner) {
                            this.angularZoneRunner = angularZoneRunner;
                        }

                        @Override
                        public void updateState() {
                            itemCount = baseItemUiService.get().getMyItemCount(itemType.getId());
                            itemLimit = gameUiControl.get().getMyLimitation4ItemType(itemType.getId());

                            if (baseItemUiService.get().isMyLevelLimitation4ItemTypeExceeded(itemType, 1)) {
                                buildLimitReached = true;
                                enabled = false;
                            } else if (baseItemUiService.get().isMyHouseSpaceExceeded(itemType, 1)) {
                                buildHouseSpaceReached = true;
                                enabled = false;
                            } else if (itemType.getPrice() > baseItemUiService.get().getResources()) {
                                buildNoMoney = true;
                                enabled = false;
                            } else {
                                enabled = true;

                            }
                        }

                        @Override
                        public void updateResources(int resources) {
                            if (itemType.getPrice() > resources && enabled) {
                                buildNoMoney = true;
                                enabled = false;
                            } else if (itemType.getPrice() <= resources && !enabled) {
                                updateState();
                            }
                        }
                    };
                    buildupItemInfo.imageUrl = CommonUrl.getImageServiceUrlSafe(itemType.getThumbnail());
                    buildupItemInfo.itemTypeId = itemType.getId();
                    buildupItemInfo.itemTypeName = itemType.getName();
                    buildupItemInfo.price = itemType.getPrice();
                    buildupItemInfo.updateState();
                    buildupItemCockpits.add(buildupItemInfo);
                    return buildupItemInfo;
                }).toArray(BuildupItemCockpit[]::new);
    }

    private ItemContainerCockpit createItemContainerInfo(BaseItemType baseItemType, Group selectedGroup) {
        if (baseItemType.getItemContainerType() != null) {
            SyncBaseItemMonitor transporter = selectedGroup.getSyncBaseItemsMonitors().stream().findFirst().orElseThrow(RuntimeException::new);
            ItemContainerCockpit itemContainerCockpit = new ItemContainerCockpit() {
                @Override
                public void setAngularZoneRunner(AngularZoneRunner angularZoneRunner) {
                    this.angularZoneRunner = angularZoneRunner;
                }

                @Override
                public void onUnload() {
                    try {
                        int baseItemTypeId = transporter.getSyncBaseItemState().getContainingItemTypeIds()[0];
                        BaseItemPlacerConfig baseItemPlacerConfig = new BaseItemPlacerConfig().baseItemCount(1).baseItemTypeId(baseItemTypeId);
                        baseItemPlacerService.activate(baseItemPlacerConfig, true, decimalPositions -> {
                            gameEngineControl.get().unloadContainerCmd(transporter.getSyncItemId(), CollectionUtils.getFirst(decimalPositions));
                        });
                    } catch (Exception e) {
                        logger.log(Level.WARNING, e.getMessage(), e);
                    }
                }
            };
            transporter.setContainingChangeListener(syncItemMonitor -> {
                if (itemContainerCockpit.angularZoneRunner == null) {
                    logger.info("temContainerCockpit.angularZoneRunner == null in ItemCockpitService.createItemContainerInfo()");
                } else {
                    itemContainerCockpit.angularZoneRunner.runInAngularZone(() ->
                            // Do not replace by methode reference. GWT can not handle that.
                            itemContainerCockpit.count = transporter.getSyncBaseItemState().getContainingItemTypeIds() != null ? transporter.getSyncBaseItemState().getContainingItemTypeIds().length : 0
                    );
                }
            });

            itemContainerCockpit.count = transporter.getSyncBaseItemState().getContainingItemTypeIds() != null ? transporter.getSyncBaseItemState().getContainingItemTypeIds().length : 0;
            return itemContainerCockpit;
        } else {
            return null;
        }
    }

    private OwnItemCockpit createSimpleOwnItemCockpit(BaseItemType baseItemType, Collection<SyncBaseItemSimpleDto> items) {
        OwnItemCockpit ownInfoPanel = new OwnItemCockpit();
        ownInfoPanel.imageUrl = CommonUrl.getImageServiceUrlSafe(baseItemType.getThumbnail());
        ownInfoPanel.itemTypeName = baseItemType.getName();
        ownInfoPanel.itemTypeDescr = baseItemType.getDescription();
        if (!gameUiControl.get().isSellSuppressed()) {
            ownInfoPanel.sellHandler = () -> {
                gameEngineControl.get().sellItems(items);
            };
        }
        return ownInfoPanel;
    }

    // This method is may not called enough. Only called on Level change and houseSpace and usedHouseSpace changed
    // If an item is created  and item of a different item type is killed, this is method not called
    public void onStateChanged() {
        buildupItemCockpits.forEach(buildupItemCockpit -> {
                    if (buildupItemCockpit.angularZoneRunner == null) {
                        logger.info("buildupItemCockpit.angularZoneRunner == null in ItemCockpitService.onStateChanged()");
                    } else {
                        buildupItemCockpit.angularZoneRunner.runInAngularZone(() ->
                                // Do not replace by methode reference. GWT can not handle that.
                                buildupItemCockpit.updateState()
                        );
                    }
                }
        );
    }

    public void onResourcesChanged(int resources) {
        buildupItemCockpits.forEach(buildupItemCockpit -> {
            if (buildupItemCockpit.angularZoneRunner == null) {
                logger.info("buildupItemCockpit.angularZoneRunner == null in ItemCockpitService.onResourcesChanged()");
            } else {
                buildupItemCockpit.angularZoneRunner.runInAngularZone(() -> buildupItemCockpit.updateResources(resources));
            }
        });
    }
}

File: razarion-frontend/src/app/game/cockpit/item/item-cockpit.service.ts
import { Injectable } from '@angular/core';
import {
  BaseItemType,
  BaseItemUiService,
  Character,
  Diplomacy,
  GameUiControl,
  ItemCockpitBridge,
  ItemTypeService,
  PlayerBaseDto
} from '../../../gwtangular/GwtAngularFacade';
import { GwtAngularService } from '../../../gwtangular/GwtAngularService';
import { SelectionService } from '../../selection.service';
import { CockpitDisplayService } from '../cockpit-display.service';

// --- View-Model Interfaces ---

export interface OwnItemCockpitModel {
  imageUrl: string;
  itemTypeName: string;
  itemTypeDescr: string;
  buildupItems: BuildupItemModel[] | null;
  containerCount: number | null;
  containerId: number | null;
  canSell: boolean;
}

export interface BuildupItemModel {
  imageUrl: string;
  itemTypeId: number;
  itemTypeName: string;
  price: number;
  itemCount: number;
  itemLimit: number;
  enabled: boolean;
  buildLimitReached: boolean;
  buildHouseSpaceReached: boolean;
  buildNoMoney: boolean;
}

export interface OwnMultipleModel {
  ownItemCockpit: OwnItemCockpitModel;
  count: number;
  baseItemTypeId: number;
}

export interface OtherItemCockpitModel {
  id: number;
  imageUrl: string;
  itemTypeName: string;
  itemTypeDescr: string;
  baseId?: number;
  baseName: string;
  friend: boolean;
  bot: boolean;
  resource: boolean;
  box: boolean;
}

function getImageServiceUrl(thumbnailId: number | null): string {
  return thumbnailId != null ? `/rest/image/${thumbnailId}` : '';
}

@Injectable({ providedIn: 'root' })
export class ItemCockpitService {
  ownItemCockpit: OwnItemCockpitModel | null = null;
  ownMultipleItems: OwnMultipleModel[] | null = null;
  otherItemCockpit: OtherItemCockpitModel | null = null;
  count: number = 0;
  private initialized = false;

  constructor(
    private selectionService: SelectionService,
    private gwtAngularService: GwtAngularService,
    private cockpitDisplayService: CockpitDisplayService
  ) {
    selectionService.addSelectionListener(() => this.onSelectionChanged());
  }

  private get itemTypeService(): ItemTypeService {
    return this.gwtAngularService.gwtAngularFacade.itemTypeService;
  }

  private get baseItemUiService(): BaseItemUiService {
    return this.gwtAngularService.gwtAngularFacade.baseItemUiService;
  }

  private get gameUiControl(): GameUiControl {
    return this.gwtAngularService.gwtAngularFacade.gameUiControl;
  }

  private get itemCockpitBridge(): ItemCockpitBridge {
    return this.gwtAngularService.gwtAngularFacade.itemCockpitBridge;
  }

  init(): void {
    this.initialized = true;
    this.itemCockpitBridge.setCockpitStateCallback(() => this.onStateChanged());
  }

  private onSelectionChanged(): void {
    // Lazy init: bridge becomes available after WASM loads
    if (!this.gwtAngularService.gwtAngularFacade.itemCockpitBridge) return;
    if (!this.initialized) {
      this.init();
    }
    // Unwatch container from previous selection
    this.itemCockpitBridge.unwatchContainerCount();

    const selectedItems = this.selectionService.getSelectedOwnItems();
    if (selectedItems.length > 0) {
      // Own selection
      const groupedByType = new Map<number, typeof selectedItems>();
      for (const item of selectedItems) {
        const typeId = item.getBaseItemType().getId();
        let group = groupedByType.get(typeId);
        if (!group) {
          group = [];
          groupedByType.set(typeId, group);
        }
        group.push(item);
      }

      if (groupedByType.size === 1) {
        // Single type selected
        const [typeId, items] = groupedByType.entries().next().value!;
        const baseItemType = this.itemTypeService.getBaseItemTypeAngular(typeId);
        this.ownItemCockpit = this.createOwnItemCockpit(baseItemType, items);
        this.ownMultipleItems = null;
        this.otherItemCockpit = null;
        this.count = items.length;
      } else {
        // Multiple types selected
        this.ownItemCockpit = null;
        this.ownMultipleItems = this.createOwnMultipleInfo(groupedByType);
        this.otherItemCockpit = null;
        this.count = selectedItems.length;
      }
      this.cockpitDisplayService.showItemCockpit = true;
    } else if (this.selectionService.getSelectedOtherId() != null) {
      // Other selection
      this.ownItemCockpit = null;
      this.ownMultipleItems = null;
      this.otherItemCockpit = this.createOtherItemCockpit();
      this.count = 1;
      this.cockpitDisplayService.showItemCockpit = true;
    } else {
      // No selection
      this.ownItemCockpit = null;
      this.ownMultipleItems = null;
      this.otherItemCockpit = null;
      this.count = 0;
      this.cockpitDisplayService.showItemCockpit = false;
    }
  }

  private createOwnItemCockpit(baseItemType: BaseItemType, items: { getId(): number }[]): OwnItemCockpitModel {
    const canSell = !this.gameUiControl.isSellSuppressed();

    const model: OwnItemCockpitModel = {
      imageUrl: getImageServiceUrl(baseItemType.getThumbnail()),
      itemTypeName: baseItemType.getName(),
      itemTypeDescr: baseItemType.getDescription(),
      buildupItems: this.createBuildupItems(baseItemType),
      containerCount: null,
      containerId: null,
      canSell
    };

    // Container info
    if (baseItemType.getItemContainerType() != null && items.length === 1) {
      const containerId = items[0].getId();
      model.containerId = containerId;
      model.containerCount = 0; // Initial; will be updated by watch
      this.itemCockpitBridge.watchContainerCount(containerId, (count: number) => {
        if (this.ownItemCockpit && this.ownItemCockpit.containerId === containerId) {
          this.ownItemCockpit.containerCount = count;
        }
      });
    }

    return model;
  }

  private createBuildupItems(baseItemType: BaseItemType): BuildupItemModel[] | null {
    let ableToBuildIds: number[] | null = null;

    if (baseItemType.getBuilderType() != null) {
      ableToBuildIds = baseItemType.getBuilderType()!.getAbleToBuildIds();
    } else if (baseItemType.getFactoryType() != null) {
      ableToBuildIds = baseItemType.getFactoryType()!.getAbleToBuildIds();
    }

    if (!ableToBuildIds || ableToBuildIds.length === 0) {
      return null;
    }

    const planetConfig = this.gameUiControl.getPlanetConfig();
    const models: BuildupItemModel[] = [];

    for (const itemTypeId of ableToBuildIds) {
      if (planetConfig.imitation4ItemType(itemTypeId) <= 0) {
        continue;
      }

      const toBuild = this.itemTypeService.getBaseItemTypeAngular(itemTypeId);
      const model = this.createBuildupItemModel(toBuild);
      models.push(model);
    }

    return models.length > 0 ? models : null;
  }

  private createBuildupItemModel(toBuild: BaseItemType): BuildupItemModel {
    const itemTypeId = toBuild.getId();
    const itemCount = this.baseItemUiService.getMyItemCount(itemTypeId);
    const itemLimit = this.gameUiControl.getMyLimitation4ItemType(itemTypeId);
    const price = toBuild.getPrice();
    const resources = this.baseItemUiService.getResources();
    const usedHouseSpace = this.baseItemUiService.getUsedHouseSpace();
    const houseSpace = this.baseItemUiService.getHouseSpace();
    const planetHouseSpace = this.gameUiControl.getPlanetConfig().getHouseSpace();
    const totalHouseSpace = houseSpace + planetHouseSpace;

    let enabled = true;
    let buildLimitReached = false;
    let buildHouseSpaceReached = false;
    let buildNoMoney = false;

    if (itemCount + 1 > itemLimit) {
      buildLimitReached = true;
      enabled = false;
    } else if (usedHouseSpace + toBuild.getConsumingHouseSpace() > totalHouseSpace) {
      buildHouseSpaceReached = true;
      enabled = false;
    } else if (price > resources) {
      buildNoMoney = true;
      enabled = false;
    }

    return {
      imageUrl: getImageServiceUrl(toBuild.getThumbnail()),
      itemTypeId,
      itemTypeName: toBuild.getName(),
      price,
      itemCount,
      itemLimit,
      enabled,
      buildLimitReached,
      buildHouseSpaceReached,
      buildNoMoney
    };
  }

  private createOwnMultipleInfo(groupedByType: Map<number, { getId(): number, getBaseItemType(): BaseItemType }[]>): OwnMultipleModel[] {
    const models: OwnMultipleModel[] = [];
    for (const [typeId, items] of groupedByType) {
      const baseItemType = this.itemTypeService.getBaseItemTypeAngular(typeId);
      const canSell = !this.gameUiControl.isSellSuppressed();
      models.push({
        ownItemCockpit: {
          imageUrl: getImageServiceUrl(baseItemType.getThumbnail()),
          itemTypeName: baseItemType.getName(),
          itemTypeDescr: baseItemType.getDescription(),
          buildupItems: null,
          containerCount: null,
          containerId: null,
          canSell
        },
        count: items.length,
        baseItemTypeId: typeId
      });
    }
    return models;
  }

  private createOtherItemCockpit(): OtherItemCockpitModel {
    const id = this.selectionService.getSelectedOtherId()!;
    const diplomacy = this.selectionService.getSelectedOtherDiplomacy()!;
    const itemTypeId = this.selectionService.getSelectedOtherItemTypeId();
    const baseId = this.selectionService.getSelectedOtherBaseId();

    let imageUrl = '';
    let itemTypeName = '';
    let itemTypeDescr = '';
    let baseName = '';
    let friend = false;
    let bot = false;
    let resource = false;
    let box = false;

    if (itemTypeId != null) {
      if (diplomacy === Diplomacy.RESOURCE) {
        const resourceType = this.itemTypeService.getResourceItemTypeAngular(itemTypeId);
        imageUrl = getImageServiceUrl(resourceType.getThumbnail());
        itemTypeName = resourceType.getName();
        itemTypeDescr = resourceType.getDescription();
        resource = true;
      } else if (diplomacy === Diplomacy.BOX) {
        // Box items don't have a specific TS type service method yet
        imageUrl = '';
        itemTypeName = 'Box';
        itemTypeDescr = '';
        box = true;
      } else {
        // Base item (FRIEND, ENEMY)
        const baseItemType = this.itemTypeService.getBaseItemTypeAngular(itemTypeId);
        imageUrl = getImageServiceUrl(baseItemType.getThumbnail());
        itemTypeName = baseItemType.getName();
        itemTypeDescr = baseItemType.getDescription();

        if (baseId != null) {
          const bases = this.baseItemUiService.getBases();
          const playerBase = bases.find((b: PlayerBaseDto) => b.getBaseId() === baseId);
          if (playerBase) {
            const character = playerBase.getCharacter();
            if (character === Character.HUMAN) {
              const name = playerBase.getName();
              if (!name || name.trim() === '') {
                baseName = 'Unnamed user';
              } else {
                baseName = name;
              }
              friend = diplomacy === Diplomacy.FRIEND;
            } else if (character === Character.BOT) {
              baseName = playerBase.getName();
              bot = true;
            } else if (character === Character.BOT_NCP) {
              baseName = playerBase.getName();
              friend = true;
              bot = true;
            }
          }
        }
      }
    }

    return {
      id,
      imageUrl,
      itemTypeName,
      itemTypeDescr,
      baseId: baseId ?? undefined,
      baseName,
      friend,
      bot,
      resource,
      box
    };
  }

  /** Called from Java when item count, house space, or resources change */
  private onStateChanged(): void {
    if (this.ownItemCockpit?.buildupItems) {
      this.refreshBuildupState(this.ownItemCockpit.buildupItems);
    }
    if (this.ownMultipleItems) {
      for (const multi of this.ownMultipleItems) {
        if (multi.ownItemCockpit.buildupItems) {
          this.refreshBuildupState(multi.ownItemCockpit.buildupItems);
        }
      }
    }
  }

  private refreshBuildupState(buildupItems: BuildupItemModel[]): void {
    const resources = this.baseItemUiService.getResources();
    const usedHouseSpace = this.baseItemUiService.getUsedHouseSpace();
    const houseSpace = this.baseItemUiService.getHouseSpace();
    const planetHouseSpace = this.gameUiControl.getPlanetConfig().getHouseSpace();
    const totalHouseSpace = houseSpace + planetHouseSpace;

    for (const item of buildupItems) {
      const toBuild = this.itemTypeService.getBaseItemTypeAngular(item.itemTypeId);
      item.itemCount = this.baseItemUiService.getMyItemCount(item.itemTypeId);
      item.itemLimit = this.gameUiControl.getMyLimitation4ItemType(item.itemTypeId);

      item.buildLimitReached = false;
      item.buildHouseSpaceReached = false;
      item.buildNoMoney = false;
      item.enabled = true;

      if (item.itemCount + 1 > item.itemLimit) {
        item.buildLimitReached = true;
        item.enabled = false;
      } else if (usedHouseSpace + toBuild.getConsumingHouseSpace() > totalHouseSpace) {
        item.buildHouseSpaceReached = true;
        item.enabled = false;
      } else if (item.price > resources) {
        item.buildNoMoney = true;
        item.enabled = false;
      }
    }
  }

  // --- Command methods called from the component ---

  onBuild(itemTypeId: number): void {
    const selectedItems = this.selectionService.getSelectedOwnItems();
    if (selectedItems.length === 0) return;

    const firstItem = selectedItems[0];
    const baseItemType = firstItem.getBaseItemType();

    if (baseItemType.getBuilderType() != null) {
      this.itemCockpitBridge.requestBuild(firstItem.getId(), itemTypeId);
    } else if (baseItemType.getFactoryType() != null) {
      const factoryIds = selectedItems.map(item => item.getId());
      this.itemCockpitBridge.requestFabricate(factoryIds, itemTypeId);
    }
  }

  onSell(): void {
    const itemIds = this.selectionService.getSelectedOwnItemIds();
    if (itemIds.length > 0) {
      this.itemCockpitBridge.sellItems(itemIds);
    }
  }

  onUnload(containerId: number): void {
    this.itemCockpitBridge.requestUnload(containerId);
  }

  keepOnlyOfType(baseItemTypeId: number): void {
    this.selectionService.keepOnlyOfType(baseItemTypeId);
  }
}
